### PR TITLE
Add Fake Cross Ref Mode, migrate sparse to it

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85417
* __->__ #85382

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -466,11 +466,19 @@ def create_aot_dispatcher_function(
     # NB: don't bother setting allow_fallback_kernels; this should not actually
     # be configurable in fake tensor, we should automatically do the right
     # thing
+    if config.debug_fake_cross_ref:
+        # This is a little messy but TorchDynamo directly changes `use_fake_tensor`
+        # so it's not enough for user to change the config manually
+        # TODO: have TorchDynamo read in `use_fake_tensor` from os environ /
+        # coordinate flags
+        config.use_fake_tensor = False
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:474]: This doesn't look kosher. Why not use the property trick to make use_fake_tensor a computed value?

File: torch/_subclasses/fake_utils.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            func
+            not in (
+                aten.lift_fresh.default,
+                aten.lift_fresh_copy.default,
+                aten.set_.source_Storage_storage_offset,
+            )
+            and not self.ignore_op_fn(func)
+            and torch.Tag.dynamic_output_shape not in func.tags  # type: ignore[attr-defined]
+            and torch.Tag.inplace_view not in func.tags  # type: ignore[attr-defined]
+            and torch.Tag.data_dependent_output not in func.tags  # type: ignore[attr-defined]
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:46]: The ignore attr-defineds shouldn't be necessary 🤔

File: torch/_subclasses/fake_utils.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        ):
+            try:
+                with FakeTensorMode() as fake_mode:
+                    fake_args, fake_kwargs = pytree.tree_map_only(
+                        torch.Tensor, fake_mode.from_tensor, (args, kwargs)
+                    )
+                    fake_r = func(*fake_args, **fake_kwargs)
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:53]: It's probably a good idea to suppress warnings here too

File: torch/_subclasses/fake_utils.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        r = func(*args, **kwargs)
+        if fake_r is not None:
+            for r_out, fake_out in zip(tree_flatten(r)[0], tree_flatten(fake_r)[0]):
+                r_ten = isinstance(r_out, torch.Tensor)
+                assert r_ten == isinstance(fake_out, torch.Tensor)
+                if r_ten:
+                    try:
+                        torch._prims.utils.compare_tensor_meta(
+                            r_out, fake_out, check_strides=self.check_strides
+                        )
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:66]: IMO the logic in test_meta.py for this is better. See `assert_ref_meta_equal`

**Review by ezyang — APPROVED** [review]
das the good stuff

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (eellison) [reply functorch/_src/aot_autograd.py:474]: What is the property trick ?

Comment (ezyang) [reply functorch/_src/aot_autograd.py:474]: check the bottom of torchdynamo/config.py. you'd customize setattr

**Comment by eellison** [comment]
Going to address comments in the follow up PR, CI green 🚀

**Pull request closed without merging**